subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1270 @@
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+
+	if (usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
+		return desc->if1_1_hs_in_ep.bEndpointAddress;
+	}
+
+	return desc->if1_1_in_ep.bEndpointAddress;
+}   /* cdc_ncm_get_bulk_in */
+
+
+
+static uint16_t cdc_ncm_get_bulk_in_mps(struct usbd_class_data *const c_data)
+{
+	struct usbd_context *uds_ctx = usbd_class_get_ctx(c_data);
+
+	if (usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
+		return 512U;
+	}
+
+	return 64U;
+}   /* cdc_ncm_get_bulk_in_mps */
+
+
+
+static uint8_t cdc_ncm_get_bulk_out(struct usbd_class_data *const c_data)
+{
+	struct usbd_context *uds_ctx = usbd_class_get_ctx(c_data);
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+
+	if (usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
+		return desc->if1_1_hs_out_ep.bEndpointAddress;
+	}
+
+	return desc->if1_1_out_ep.bEndpointAddress;
+}   /* cdc_ncm_get_bulk_out */
+
+
+
+static struct net_buf *cdc_ncm_buf_alloc(const uint8_t ep)
+{
+	struct net_buf *buf = NULL;
+	struct udc_buf_info *bi;
+
+	buf = net_buf_alloc(&cdc_ncm_ep_pool, K_NO_WAIT);
+	if (!buf) {
+		return NULL;
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
+	bi->ep = ep;
+
+	return buf;
+}   /* cdc_ncm_buf_alloc */
+
+
+
+static int cdc_ncm_out_start(struct usbd_class_data *const c_data)
+/**
+ * Initiate reception
+ */
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct net_buf *buf;
+	uint8_t ep;
+	int ret;
+
+	LOG_DBG("%ld", data->state);
+	if (!atomic_test_bit(&data->state, CDC_NCM_CLASS_ENABLED)) {
+		return -EACCES;
+	}
+
+	if (atomic_test_and_set_bit(&data->state, CDC_NCM_OUT_ENGAGED)) {
+		return -EBUSY;
+	}
+
+	ep = cdc_ncm_get_bulk_out(c_data);
+	buf = cdc_ncm_buf_alloc(ep);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	ret = usbd_ep_enqueue(c_data, buf);
+	if (ret) {
+		LOG_ERR("Failed to enqueue net_buf for 0x%02x", ep);
+		net_buf_unref(buf);
+	}
+
+	return  ret;
+}   /* cdc_ncm_out_start */
+
+
+
+static bool _cdc_ncm_frame_ok(struct cdc_ncm_eth_data *data, struct net_buf *const buf)
+/**
+ * check a received NTB
+ */
+{
+	const union recv_ntb_t *ntb = (union recv_ntb_t *)buf->data;
+	const struct nth16_t *nth16 = &(ntb->nth);
+	const uint16_t len = buf->len;
+
+	LOG_DBG("%p, %d", ntb, (int)len);
+
+	/*
+	 * check header
+	 */
+	if (len < sizeof(ntb->nth)) {
+		LOG_ERR("  ill length: %d", len);
+		return false;
+	}
+	if (sys_le16_to_cpu(nth16->wHeaderLength) != sizeof(struct nth16_t)) {
+		LOG_ERR("  ill nth16 length: %d", sys_le16_to_cpu(nth16->wHeaderLength));
+		return false;
+	}
+	if (sys_le32_to_cpu(nth16->dwSignature) != NTH16_SIGNATURE) {
+		LOG_ERR("  ill signature: 0x%08x",
+				(unsigned int)sys_le32_to_cpu(nth16->dwSignature));
+		return false;
+	}
+	if (len < sizeof(struct nth16_t) + sizeof(struct ndp16_t)
+			+ 2*sizeof(struct ndp16_datagram_t)) {
```

Comment:
What is `sizeof(struct nth16_t) + sizeof(struct ndp16_t)	+ 2*sizeof(struct ndp16_datagram_t)` ?